File: websocket.js
//以下代码除console.log以外，其他代码随意修改都可能出现bug
var socket
function dl() {//当输入完初始信息后调用此函数连接服务器
    socket = new WebSocket("ws://127.0.0.1:5418")
    socket.addEventListener("open", function (event) {//第一次连接
        senddata = {
            "type": "open",
            "img": "img/a.JPG",//头像（可自行修改）
            "name": document.getElementById("name").value,//名字
            "epithet": document.getElementById("epithet").value//称号
        }
        fasong(senddata)//发送初始化信息
        document.getElementById("dldiv").remove()//删掉初始化的按钮及输入框，防止出现问题
    })

    socket.addEventListener("message", function (event) {//当收到服务器信息或转发时
        var data = JSON.parse(event.data)//字符串转json
        console.log(data)
    })

    socket.addEventListener("close", function (event) {
        console.log("close")
    })

    socket.addEventListener("error", function (event) {
        console.log("server error")
    })
}

function fasong(json) {//发送至服务器
    var data = JSON.stringify(json)//json转字符串
    socket.send(data)//发送
}

onbeforeunload = function () {//当刷新或关闭网页时向服务器发送退出信息
    fasong({"type":"close","text":"bye"})
}

function send() {//发送聊天文字
    senddata = {
        "type": "send",
        "text": document.getElementById('fasong').value
    }
    fasong(senddata)
}

File: websocket.py
import asyncio
import websockets
import json

Users = {}


async def chat(websocket):#聊天主模块
    global Users#全局变量，用来存储连接的前端

    async for message in websocket:
        data = json.loads(message)#发来的字符串转json
        if data["type"] == "open":#初次连接
            Users.update({websocket: {"img": data["img"], "name": data["name"], "epithet": data["epithet"]}})#想字典里存储{websocket连接:{"头像":"","名字":"","称号":""}}
            you = {#返回的数据
                "type": "open"
            }
            await websocket.send(json.dumps(you))#将连接成功的提示返回回去

        if data["type"] == "send":#客户端发送聊天信息
            user = Users[websocket]#获取此客户端信息
            senddata = {#要转发给其他人的json
                "type": "send",#类型
                "img": user["img"],#此用户头像
                "name": user["name"],#此用户名称
                "epithet": user["epithet"],#此用户称号
                "text": data["text"]#此用户发的文字
            }
            for key in Users.keys():#例遍在线的人的websocket
                await asyncio.wait([key.send(json.dumps(senddata))])#逐一转发

        if data["type"] == "close":#当有用户退出时
            del Users[websocket]#将他从users中删除，下次转发将不会再发给他，否则会报错


async def main():#入口
    async with websockets.serve(chat, "127.0.0.1", 5418):
        await asyncio.Future()

#asyncio.run(main())